### PR TITLE
webUI tests for admin external storage page

### PR DESCRIPTION
**Pull request by paurakhsharma**
## Description
This PR webUI tests for files_external.

Related issue:  #32852

## How Has This Been Tested?
🤖

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Acceptance test

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: tests/acceptance/features/lib/AdminStorageSettingsPage.php [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+					"xpath", \sprintf($this->applibableUserDeleteXpath, $i)
+				);
+				$applibableUserDelete->click();
+				break;
+			}
+			throw new \Exception(
+				__METHOD__ . " could not find $user to remove from applicable list"
+			);
+		}
+		$this->waitUntillSuccessOrFailureSymbol();
```
Comment (skshetry) [inline tests/acceptance/features/lib/AdminStorageSettingsPage.php]:
```suggestion
		$this->waitUntilSuccessOrFailureSymbol();
```

File: tests/acceptance/features/lib/AdminStorageSettingsPage.php [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+	}
+
+	/**
+	 * Wait untill sucess or failure symbol to appear
+	 *
+	 * @param integer $timeout time in ms
+	 *
+	 * @return void
+	 */
+	public function waitUntillSuccessOrFailureSymbol(
```
Comment (skshetry) [inline tests/acceptance/features/lib/AdminStorageSettingsPage.php]:
```suggestion
	public function waitUntilSuccessOrFailureSymbol(
```

**Review by skshetry — CHANGES_REQUESTED** [review]

File: tests/acceptance/features/bootstrap/WebUIAdminStorageSettingsContext.php [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+	 * @When /^the administrator (adds|removes) (user|group) "([^"]*)" (?:as|from) the applicable (?:user|group) for the last local storage mount using the webUI$/
+	 * @Given /^the administrator has (added|removed) (user|group) "([^"]*)" (?:as|from) the applicable (?:user|group) for the last local storage mount from the admin storage settings page$/
+	 *
+	 * @param string $action
+	 * @param string $userOrGroup
+	 * @param string $user
+	 *
+	 * @return void
+	 */
+	public function theAdministratorAddsUserAsTheApplicableUserForTheLastLocalStorageMountUsingTheUsingTheWebui($action, $userOrGroup, $user) {
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebUIAdminStorageSettingsContext.php]: put parameters on a new line

File: tests/acceptance/features/lib/AdminStorageSettingsPage.php [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+ */
+class AdminStorageSettingsPage extends OwncloudPage {
+
+	/**
+	 *
+	 * @var string $path
+	 */
+	protected $path = '/index.php/settings/admin?sectionid=storage';
+
+	protected $filesExternalForumXpath = "//form[@id='files_external']";
```
Comment (phil-davis) [inline tests/acceptance/features/lib/AdminStorageSettingsPage.php]: Forum or Form?

File: tests/acceptance/features/lib/AdminStorageSettingsPage.php [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+	protected $newFolderNameXpath = "//tr[@id='addMountPoint']//input[@placeholder='Folder name']";
+	protected $newBackendTypeXpath = "//tr[@id='addMountPoint']//select[@id='selectBackend']";
+	protected $newLocationXpath = "//tr[@id='addMountPoint']/preceding-sibling::tr[1]//input[@placeholder='Location']";
+	protected $statusSymbolXpath = "//tr[@id='addMountPoint']/preceding-sibling::tr[1]//td[@class='status']/span";
+
+	protected $lastMountApplicableXpath = "//tr[@id='addMountPoint']/preceding-sibling::tr[1]//td[@class='applicable']//li/input";
+	protected $userOrGroupListXpath = "//div[@id='select2-drop']//li[contains(@class, 'select2-result')]";
+
+	protected $applicableUsersListXpath = "//tr[@id='addMountPoint']/preceding-sibling::tr[1]//li[@class='select2-search-choice']";
+	protected $applibableUserXpath = "//tr[@id='addMountPoint']/preceding-sibling::tr[1]//li[@class='select2-search-choice'][%s]//span";
```
Comment (phil-davis) [inline tests/acceptance/features/lib/AdminStorageSettingsPage.php]: s/applibableUser/applicableUser/

File: tests/acceptance/features/lib/AdminStorageSettingsPage.php [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+	protected $applibableUserDeleteXpath = "//tr[@id='addMountPoint']/preceding-sibling::tr[1]//li[@class='select2-search-choice'][%s]//a";
```
Comment (phil-davis) [inline tests/acceptance/features/lib/AdminStorageSettingsPage.php]: s/applibableUser/applicableUser/

File: tests/acceptance/features/bootstrap/WebUIAdminStorageSettingsContext.php [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+	private $webUIGeneralContext;
+
+	/**
+	 *
+	 * @var FeatureContext
+	 */
+	private $featureContext;
+
+	/**
+	 * WebUIAdminStorageSettingsContext constructor.
```
[the remark this answers is not in the document]
Comment (paurakhsharma) [reply tests/acceptance/features/bootstrap/WebUIAdminStorageSettingsContext.php]: That's a great suggestion but that's the way docstring is written in all webUI contexts.

File: tests/acceptance/features/lib/AdminStorageSettingsPage.php [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+		$newLocation = $this->find("xpath", $this->newLocationXpath);
+		$this->assertElementNotNull(
+			$newLocation,
+			__METHOD__ .
+			" xpath $this->newLocationXpath " .
+			"could not find input field for local storage mount location"
+		);
+		$newLocation->setValue($dirLocation);
+		$newLocation->blur();
+		$this->waitUntillSuccessOrFailureSymbol();
```
Comment (phil-davis) [inline tests/acceptance/features/lib/AdminStorageSettingsPage.php]: s/Untill/Until

File: tests/acceptance/features/lib/AdminStorageSettingsPage.php (hunk 1 above)
Comment (paurakhsharma) [reply tests/acceptance/features/lib/AdminStorageSettingsPage.php]: Oh, there are no two l's in untill 😄

File: tests/acceptance/features/lib/AdminStorageSettingsPage.php [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+	 * @throws \Exception
+	 */
+	public function removeApplicableFromLastMount($session, $user) {
+		$applicableUsersList = $this->findAll("xpath", $this->applicableUsersListXpath);
+		for ($i = 1; $i <= \count($applicableUsersList); $i++) {
+			$applicableUser = $this->find(
+				"xpath", \sprintf($this->applibableUserXpath, $i)
+			);
+			if ($this->getTrimmedText($applicableUser) === \strtolower($user)) {
+				$applibableUserDelete = $this->find(
```
Comment (phil-davis) [inline tests/acceptance/features/lib/AdminStorageSettingsPage.php]: s/applibable/applicable

File: tests/acceptance/features/lib/AdminStorageSettingsPage.php [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+			}
+			throw new \Exception(
+				__METHOD__ . " could not find $user to remove from applicable list"
+			);
+		}
+		$this->waitUntillSuccessOrFailureSymbol();
+	}
+
+	/**
+	 * Wait untill sucess or failure symbol to appear
```
Comment (phil-davis) [inline tests/acceptance/features/lib/AdminStorageSettingsPage.php]: s/untill sucess/until success 

Wait until success or failure symbol appears

File: tests/acceptance/features/lib/AdminStorageSettingsPage.php (hunk 2 above)
Comment (phil-davis) [inline tests/acceptance/features/lib/AdminStorageSettingsPage.php]: s/Untill/Until

File: tests/acceptance/features/webUIAdminSettings/adminStorageSettings.feature [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+
+  Scenario: administrator creates a local storage mount
+    Given user "user0" has been created with default attributes
+    And the administrator has browsed to the admin storage settings page
+    And the administrator has enabled the external storage
+    When the administrator creates the local storage mount "local_storage1" using the webUI
+    And the user re-logs in as "user0" using the webUI
+    Then folder "local_storage1" should be listed on the webUI
+
+  Scenario: administrator assigns a applicable user to a local storage mount
```
Comment (phil-davis) [inline tests/acceptance/features/webUIAdminSettings/adminStorageSettings.feature]: s/a applicable user/an applicable user

**Review by phil-davis — COMMENTED** [review]
Looks like "a good thing". I only found typos.

File: tests/acceptance/features/lib/AdminStorageSettingsPage.php [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+				return;
+			}
+		}
+		throw new \Exception(
+			__METHOD__ . " could not find $user to remove from applicable list"
+		);
+	}
+
+	/**
+	 * Wait until sucess or failure symbol to appear
```
Comment (phil-davis) [inline tests/acceptance/features/lib/AdminStorageSettingsPage.php]: s/sucess/success

**Comment by paurakhsharma** [comment]
@phil-davis @skshetry Please review it again, thanks.

File: tests/acceptance/features/lib/AdminStorageSettingsPage.php [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+	}
+
+	/**
+	 * Wait until success or failure symbol to appear
+	 *
+	 * @param integer $timeout time in ms
+	 *
+	 * @return void
+	 */
+	public function waitUntilSuccessOrFailureSymbol(
```
Comment (skshetry) [inline tests/acceptance/features/lib/AdminStorageSettingsPage.php]:
```suggestion
	public function waitUntilSuccessOrFailureSymbolAppears(
```

**Review by skshetry — CHANGES_REQUESTED** [review]
LGTM except a issue. See above comment.

**Review by phil-davis — APPROVED** [review]
Looks good now. Saugat just has 1 suggestion of a method name.

**Review by skshetry — APPROVED** [review]

**Comment by paurakhsharma** [comment]
Backport on: #33855

**Pull request merged**